plugins/linguist-backend/migrations/20221115_init.js

```diff
@@ -32,7 +32,7 @@ exports.up = async function up(knex) {
       .comment('An insert counter to ensure ordering');
     table.uuid('id').notNullable().comment('The ID of the Linguist result');
     table
-      .text('entity_ref')
+      .string('entity_ref')
```

Comment:
The important bit is that they _will_ be immutable to existing installations so whatever changes you make need to be both backwards and forwards compatible, so to speak. Since an existing installation won't rerun the migration on their existing database, they don't mind that this changes as such, but the _code_ can't change in such a way that they rely on some _different outcome_ than the old migration did. This, of course, is a very delicate dance and ideally there should be migration tests (like we started a bit on [in the catalog backend](https://github.com/backstage/backstage/blob/cc14e1ba0ec638dbc9c07636e8db89e3cfd2e5cc/plugins/catalog-backend/src/tests/migrations.test.ts#L44)) that thoroughly test these to work even when there's existing data in the catalog.

Reply:
So in summary, I think it's safe to say that the only time we should allow migration files to change, is when we want to address actual breakages in them (these are effectively new breakages due to a new database being supported), and when we can do it in such a way that the outcome is effectively the same (for some definition of effectively)

Reply:
Agreed with all of the above. Ideally we wouldn't ever update migrations. I think there are solutions that in the future we can use to avoid this for breaks that occur when adding a new database type.

MySQL requires indexed columns to have a length. Eventually in knex these resolve to varchar(N) columns and will fail the migration with a missing length

If a customer uses MySQL as a backing store, as-is without changes, the migrations will fail. The MySQL db will be empty.  That is of-course only the case with the updated plugins listed in the pull request. Thus the changes to migrations directly.

I think knex has some plugins / adapters that transform the migrations at run-time and I believe the backstage oss team has ideas as well on how to combat these issues and will be happy to work through this in future updates

Reply:
Thanks everyone for the details, this all makes sense to me now with this additional information. 👍